### PR TITLE
Added Value.writeTo(), fixed instanceof Value test.

### PR DESCRIPTION
**Pull request by zslayton**
*Description of changes:*

**`dom.Value.writeTo(writer: Writer): void`** 

This PR adds a new method to the `Value` interface which writes that Ion value to a provided `Writer`.

```typescript
let writer = ion.makeBinaryWriter();
ion.dom.load('foo::bar::7').writeTo(writer);
writer.close();
let ionBytes = writer.getBytes();
```

An upcoming PR will add convenience methods to instantiate/close a `Writer` and return the result.

**Fixed behavior of the`instanceof dom.Value` test** 

`dom.Value` is a mixin constructor, which causes some unexpected behavior.

```typescript
class MyClass extends dom.Value(...) {}

let myClass = new MyClass();
myClass instanceof MyClass // true
myClass instanceof dom.Value // false
```

This can be addressed by overriding the `Value` class' implementation of `instanceof`, which is managed by the special static `Symbol.hasInstance` function. I have overridden the default behavior to test for the presence of a custom [Javascript `Symbol`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol) that is assigned to each class constructed by the `Value` mixin.

This provides the desired behavior and is very efficient:
```typescript
class MyClass extends dom.Value(...) {}

let myClass = new MyClass();
myClass instanceof MyClass // true
myClass instanceof dom.Value // true
```

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/dom/Value.ts [hunk 1]
```diff
@@ -130,6 +131,13 @@ export interface Value {
      * function will cast it as type `T` and return it; otherwise throws an Error.
      */
     as<T extends Value>(ionValueType: Constructor<T>): T;
+
+    /**
+     * Writes the Ion value represented by this dom.Value (including annotations) to the provided Writer,
```
[the remark this answers is not in the document]
Comment (zslayton) [reply src/dom/Value.ts]: Fixed.

File: src/dom/Value.ts [hunk 2]
```diff
@@ -144,6 +152,12 @@ export type PathElement = string | number;
  */
 export type Constructor<T = {}> = new (...args: any[]) => T;
 
+/**
+ * This Symbol is used to mark classes created by the Value mixin as belonging to the dom.Value
+ * class hierarchy. This allows the `className instanceof dom.Value` test to work as expected.
+ */
+const DOM_VALUE_SIGNET = Symbol("ion.dom.Value");
```
[the remark this answers is not in the document]
Comment (zslayton) [reply src/dom/Value.ts]: Done.

File: src/dom/Value.ts [hunk 3]
```diff
@@ -144,6 +152,12 @@ export type PathElement = string | number;
  */
 export type Constructor<T = {}> = new (...args: any[]) => T;
 
+/**
+ * This Symbol is used to mark classes created by the Value mixin as belonging to the dom.Value
+ * class hierarchy. This allows the `className instanceof dom.Value` test to work as expected.
+ */
```
[the remark this answers is not in the document]
Comment (zslayton) [reply src/dom/Value.ts]: Yep, good catch.

File: src/dom/Value.ts [hunk 4]
```diff
@@ -310,7 +343,28 @@ export namespace Value {
...
+    export function from(value: any, annotations?: string[]): Value {
+        if (value instanceof Value) {
+            if (_hasValue(annotations)) {
+                // TODO: This should probably be supported, but will require a clone(annotations?: string[])
+                //       API (or similar) to be added to Value subclasses.
+                throw new Error(
+                    "Value.from() does not support overriding the annotations on a dom.Value"
+                    + " passed as an argument."
+                );
+            }
```
[the remark this answers is not in the document]
Comment (zslayton) [reply src/dom/Value.ts]: I'm planning to open an issue for all the TODOs shortly.

File: test/dom/Value.ts [hunk 5]
```diff
@@ -10,6 +11,56 @@ function jsEqualsIon(jsValue, ionValue): boolean {
...
+        } else {
+            valueName = `${typeof value}(${value})`;
+        }
+        it(`${valueName} instanceof Value`, () => {
+            assert.equal(value instanceof Value, expected);
+        });
+    }
+}
+
+// Converts each argument in `values` to a dom.Value, writes it to an Ion stream, and then load()s the Ion stream back
```
[the remark this answers is not in the document]
Comment (zslayton) [reply test/dom/Value.ts]: 👍 Done.

File: test/dom/Value.ts [hunk 6]
```diff
@@ -10,6 +11,56 @@ function jsEqualsIon(jsValue, ionValue): boolean {
...
+            it(value + '', () => {
+                let writer = ion.makeBinaryWriter();
+                // Make a dom.Value out of the provided value if it isn't one already
+                let domValue;
+                if (value instanceof Value) {
+                    domValue = value as Value;
+                } else {
+                    domValue = Value.from(value);
+                }
+                // Serialize the dom.Value using the provided writer
```
[the remark this answers is not in the document]
Comment (zslayton) [reply test/dom/Value.ts]: ~~provided~~ newly instantiated

File: test/dom/Value.ts [hunk 7]
```diff
@@ -174,4 +225,146 @@ describe('Value', () => {
...
+            (c1, c2) => c1.every((value, index) => value === c2[index]),
+            load('{{"February"}}'),
+            load('month::{{"February"}}'),
+        );
+        domRoundTripTest(
+            'List',
+            (l1, l2) => l1.reduce(
+                (equalSoFar, value, index) =>
+                    // The dom.Value classes don't have an easy test for deep equality,
+                    // so here we only check for general structure.
```
[the remark this answers is not in the document]
Comment (zslayton) [reply test/dom/Value.ts]: Yes, I think we could use a `Value.equals(otherValue: Value)` implementation for all of the types. I'll open an issue.

File: test/dom/Value.ts [hunk 8]
```diff
@@ -174,4 +225,146 @@ describe('Value', () => {
...
+        );
+        domRoundTripTest(
+            'Struct',
+            (s1: Value, s2: Value) => s1.fields().reduce(
+                (equalSoFar, [key, value]) =>
+                    // The dom.Value classes don't have an easy test for deep equality,
+                    // so here we only check for general structure.
+                    equalSoFar
+                    && value.isNull() === s2[key].isNull()
+                    && value.getType() === s2[key].getType(),
```
[the remark this answers is not in the document]
Comment (zslayton) [reply test/dom/Value.ts]: Done/

File: test/dom/Value.ts [hunk 9]
```diff
@@ -174,4 +225,146 @@ describe('Value', () => {
...
+                    equalSoFar
+                        && value.isNull() === l2[index].isNull()
+                        && value.getType() === l2[index].getType(),
+                true
+            ),
+            [],
+            [1, 2, 3],
+            ['foo', 'bar', 'baz'],
+            [new Date(0), true, "hello", null]
+        );
```
[the remark this answers is not in the document]
Comment (zslayton) [reply test/dom/Value.ts]: Good catch, done.

**Comment by zslayton** [comment]
> is `value instanceof dom.Blob` a valid use case? if not, ok; if so, let's add tests

Added `instanceof` tests for `Blob` and `Clob`, thanks.

File: src/dom/Value.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (zslayton) [reply src/dom/Value.ts]: Opened issue #577 to track this.

File: test/dom/Value.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (zslayton) [reply test/dom/Value.ts]: Opened issue #576 to track this.

File: test/dom/Value.ts [hunk 10]
```diff
@@ -174,4 +225,164 @@ describe('Value', () => {
...
+        domRoundTripTest(
+            'S-Expression',
+            (s1, s2) => s1.reduce(
+                (equalSoFar, value, index) =>
+                    // The dom.Value classes don't have an easy test for deep equality,
+                    // so here we only check for general structure.
+                    equalSoFar
+                    && value.isNull() === s2[index].isNull()
+                    && value.getType() === s2[index].getType(),
```
Comment (pbcornell) [inline test/dom/Value.ts]: fix indent

File: src/dom/Value.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (pbcornell) [reply src/dom/Value.ts]: actually, why is it valid/useful for a caller to pass a `Value` to this method?  doing so is a no-op (or throws, with this change).  i imagine it's convenient for the recursive case, but that doesn't mean it should be a valid use case for callers

i'd suggest we throw anytime a `Value` is passed to this method, period.

**Comment by pbcornell** [comment]
> > is `value instanceof dom.Blob` a valid use case? if not, ok; if so, let's add tests
> 
> Added `instanceof` tests for `Blob` and `Clob`, thanks.

This isn't what i had in mind.  All the tests that call `instanceOfTest()` perform a `value instanceof dom.Value` check.  do we ever expect callers to check `value instanceof dom.Blob`, `value instanceof dom.String`, etc.?  if not, fine as-is.  if so, we should add tests for that usage

**Comment by zslayton** [comment]
> actually, why is it valid/useful for a caller to pass a Value to this method? doing so is a no-op (or throws, with this change). i imagine it's convenient for the recursive case, but that doesn't mean it should be a valid use case for callers.
> 
> i'd suggest we throw anytime a Value is passed to this method, period.

Functions that wish to accept both JS `any` and `dom.Value`s are pretty common, and accepting `dom.Value` reduces this:

```typescript
function(param: any) {
  let value: Value;
  if (param instanceof Value) {
    value = param as Value;
  } else {
    value = Value.from(param);
  }
  // Do stuff with `value`
}
```
to this:
```typescript
function(param: any) {
  let value: Value = Value.from(param);
  // Do stuff with `value`
}
```

The `throws` restriction is unfortunate and will be solved by adding [support for a `Value.clone()` method](https://github.com/amzn/ion-js/issues/577).

**Comment by zslayton** [comment]
> > > is `value instanceof dom.Blob` a valid use case? if not, ok; if so, let's add tests
> > 
> > 
> > Added `instanceof` tests for `Blob` and `Clob`, thanks.
> 
> This isn't what i had in mind. All the tests that call `instanceOfTest()` perform a `value instanceof dom.Value` check. do we ever expect callers to check `value instanceof dom.Blob`, `value instanceof dom.String`, etc.? if not, fine as-is. if so, we should add tests for that usage

Ah, ok. I wasn't sure which direction you meant because I happened to omit the `blob`/`clob` cases from the `instanceof` tests I had added. I'll add tests for the subclasses `instanceof` checks which were already functional before this PR.

**Comment by pbcornell** [comment]
> The `throws` restriction is unfortunate and will be solved by adding [support for a `Value.clone()` method](https://github.com/amzn/ion-js/issues/577).

It's an interesting tradeoff.  My app would work fine with `Value.from()` until I discover the magic that is annotations, then **boom**.  If the idea is that `Value.from()` will clone a `Value` and uses the specified annotations, it seems surprising (but maybe not problematic?) that sometimes it clones my `Value`, and other times it doesn't.

Just thinking this through.  I think your approach is reasonable and JS-friendly, and I don't have a better alternative to offer.

**Pull request merged**